geopandas/geodataframe.py

```diff
@@ -1422,6 +1408,35 @@ def apply(self, func, axis=0, raw=False, result_type=None, args=(), **kwargs):
 
     @property
     def _constructor(self):
+        geometry_default = getattr(self, "_geometry_column_name")
+        crs_default = getattr(self, "crs")
+
+        def _geodataframe_constructor_with_fallback(
+            data=None, index=None, crs=crs_default, geometry=geometry_default, **kwargs
+        ):
+            df = pd.DataFrame(data, index=index)
+            if geometry in df.columns:
+                geo_col = df[geometry]
+                if isinstance(geo_col, pd.DataFrame) or isinstance(
+                    geo_col.dtype, GeometryDtype
+                ):
+                    # DataFrame check is for when multiple columns are named
+                    # `geometry`. This case should error, but is passed through so that
+                    # we can have a specific error message for specific contexts,
+                    # e.g. concat, caught in __finalize__.
+
```

Comment:
the `if isinstance(geo_col, pd.DataFrame)` handling is a bit of a hack. Originally, I was throwing an error here in this case which is arguably the most logical thing to do, as __init__ won't permit this (and isn't throwing because of the sindex compatibility below). But throwing a generic error here means we can't throw a use case specific error for `pd.concat(axis=1)` (`test_concat_axis1_multiple_geodataframes` - added in #2046) as concat calls 
`return sample._constructor(new_data).__finalize__(self, method="concat")` and if we throw in `_constructor`, the error will be generic and we can't know that this was called from the context of concat - which was getting done in finalize. 

Allowing this through without throwing though could lead to some edge cases if other functions calling `_constructor` produce duplicate columns with name `geometry`, if they aren't caught in finalize/ don't even call finalize. So I don't really like that. The alternative options I see are to live with a generic error message, or perhaps we could monkey patch pd.concat, but that's unsatisfying as well.

Reply:
Another example where you can end up with duplicated geometry column names is `rename`. Something like:

```
df = geopandas.read_file(geopandas.datasets.get_path("naturalearth_lowres"))
df["geometry2"] = df["geometry"]
result = df.rename(columns={"geometry2": "geometry"})
```

Currently that "works" (because it only calls `__finalize__`, not `_constructor`), but then subsequent steps (eg printing it) can fail. 
(so not fully relevant for this discussion)

---

I agree it is nice to have the more specific error message for `concat`. 

---

How does this work on the main branch? Because `_geodataframe_constructor_with_fallback` currently does call GeoDataFrame(..) regardless of whether there are duplicated geometry columns present or not.

Reply:
So this was done in #2046 where there probably should have been a little more discussion on behaviour, as it's a little inconsistent. Currently, if `GeoDataFrame.__init__` gets called with geometry name None, and there is are multiple columns called "geometry" there is the generic error message "GeoDataFrame does not support multiple columns using the geometry column name 'geometry'." If the situation is the same, but geometry="geometry" is supplied, the same error is thrown by set_geometry.

If however, the geometry column is called something else, the call to `_constructor` in `concat` doesn't get the active geometry column name as an argument, so a geodataframe is constructed with geometry column called "geometry" (which doesn't exist), and then the metadata is applied in the finalize for concat, where the error is detected, and the specialised error message "Concat operation has resulted in multiple columns using the geometry column name 'geom'." is thrown.

So without the delayed assignment of the geometry column name below, set geometry will throw, which could be good or bad.